### DISCUSSION TITLE
Is it possible to activate bluetooth on bare-metal servers when running flatcar ?

### DISCUSSION THREAD

**Discussion by SR-G** [flatcar/Flatcar#1556] [Q&A]
Hello,

I'm running flatcar linux on bare-metal small servers ("NUC"-like, from "miniforum"), through IPXE (and with the system running in memory).

I would like on one of them to be able to activate the bluetooth stack, in order to run docker containers that would reach IOT devices through bluetooth (these containers have been tested on a regular server where for now they are running fine, so no issues about "docker" itself).

However it seems it may not be so easy with flatcar : at first glance it seems no bluetooth is considered / activated.

```
FLATCAR-SERVER-2 ~ # dmesg -T|grep -i bluetooth
FLATCAR-SERVER-2 ~ # 
```

(nothing found)

So ... is it possible ? How to achieve this ?

**Comment by ader1990** [comment]
Hello,

It is possible in Flatcar, there are two options: having the module and possible submodules in the Flatcar kernel or by building the modules in a systemd sysext extension and load it on demand.

First of all, you all need to check which Bluetooth hardware is present and what is the driver (s) required.

**Comment by SR-G** [comment]
Ok, so it's already good to know that it's possible ;)

But it seems a bit mysterious and not so easy : 
- regarding "having the module in the Flatcar kernel" ==> that means, rebuilding (= enhancing) the flatcar image in local ?
- whereas systemd sysext extension would be "external" (still with the standard image) and published through Ignition, that's it ?

Any recommendation about the simplest / better approach ?
Any idea of where to start ?

--- 

Regarding the hardware, it's an x86 mini-computer running quite standard Intel stuff, on my side - i would say it should be : 

```bash
# dmesg -T|grep -i tooth
[lun.  7 oct. 03:02:15 2024] Bluetooth: Core ver 2.22
[lun.  7 oct. 03:02:15 2024] NET: Registered PF_BLUETOOTH protocol family
[lun.  7 oct. 03:02:15 2024] Bluetooth: HCI device and connection manager initialized
[lun.  7 oct. 03:02:15 2024] Bluetooth: HCI socket layer initialized
[lun.  7 oct. 03:02:15 2024] Bluetooth: L2CAP socket layer initialized
[lun.  7 oct. 03:02:15 2024] Bluetooth: SCO socket layer initialized
[lun.  7 oct. 03:02:15 2024] Bluetooth: hci0: Legacy ROM 2.5 revision 1.0 build 3 week 17 2014
[lun.  7 oct. 03:02:15 2024] Bluetooth: hci0: Intel device is already patched. patch num: 32
[lun.  7 oct. 03:02:15 2024] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[lun.  7 oct. 03:02:15 2024] Bluetooth: BNEP filters: protocol multicast
[lun.  7 oct. 03:02:15 2024] Bluetooth: BNEP socket layer initialized
[lun.  7 oct. 03:02:15 2024] Bluetooth: MGMT ver 1.22
```

```bash
# lsmod | grep "btusb"
btusb                  86016  0
btrtl                  32768  1 btusb
btintel                65536  1 btusb
btbcm                  24576  1 btusb
btmtk                  12288  1 btusb
bluetooth            1105920  27 btrtl,btmtk,btintel,btbcm,bnep,btusb
```

```bash
# lsusb |grep Bluetooth
Bus 001 Device 004: ID 8087:0a2a Intel Corp. Bluetooth wireless interface
```

So this => https://linux-hardware.org/?id=usb:8087-0a2a

**Comment by ader1990** [comment]
Hello,

At this moment, as you also might need other drivers and more in the future, I would suggest to change and build Flatcar locally to try it out faster. Also, as some drivers are really pin-point solutions to specific hardware, and the main Flatcar image should be as generic and small as possible, thus limiting the merge-ability of such changes.

This is a good starting point:
https://www.flatcar.org/docs/latest/reference/developer-guides/sdk-modifying-flatcar/
Recently, there is a quite alot of gotchas / how-to-s on building Flatcar bundled in this thread: https://github.com/flatcar/Flatcar/issues/1550

Also, here is an example of what a Linux kernel config PR looks like:
https://github.com/flatcar/scripts/pull/2349

Thanks.
